Meta/check-headers-acyclic.py

```diff
@@ -0,0 +1,308 @@
+#!/usr/bin/env python3
+
+import os
+import re
+import subprocess
+import sys
+
+ANY_INCLUDE_RE = re.compile('^.*#[ \t]*include(?![A-Za-z]).*\n?$')
+SYSTEM_INCLUDE_RE = re.compile('^.*#[ \t]*include *<([^>]+)> *(?:/[*/].*)?\n?$')
+LOCAL_INCLUDE_RE = re.compile('^.*#[ \t]*include *"([^>]+)" *(?:/[*/].*)?\n?$')
+SHOW_TIMING = False
+
+# These are already very extensive. The only thing missing are the Toolchain includes.
+INCLUDE_LOOKUP_PATHS_FORMAT = [
+    '',  # For things that specify the full path, e.g. <AK/String.h>
+    'Userland/Libraries/',
+    'Userland/Libraries/LibC/',
+    'Userland/Libraries/LibCrypt/',
+    'Userland/Libraries/LibSystem/',
+    'Userland/Services/',
+    'Userland/',
+    'Build/{}/',
+    'Build/{}/Userland/Services/',
+    'Build/{}/Userland/Libraries/',
+    'Build/{}/Userland/',
+]
+
+# These system headers are too hard to find, and they're irrelevant for the purposes of this script anyway.
+KNOWN_GOOD_SYSTEM_HEADERS = {
+    'bits/endian.h',
+    'cxxabi.h',
+    'initializer_list',
+    'libkern/OSByteOrder.h',
+    'machine/endian.h',
+    'malloc/malloc.h',
+    'new',
+    'sanitizer/asan_interface.h',
+    'stdbool.h',
+    'sys/random.h',
+    'typeinfo',
+    'utility',
+    'xmmintrin.h',
+}
+
+
+def get_headers_here():
+    # FIXME: Really test *ALL* other headers!
+    result = subprocess.run(['git', 'ls-files', '*.h'], check=True, capture_output=True, text=True)
+    assert result.stderr == ''
+    output = result.stdout.split('\n')
+    assert output[-1] == ''  # Trailing newline
+    assert len(output) > 500, 'There should be well over a thousand headers, not only {}?!'.format(len(output))
+    return output[:-1]
+
+
+def find_system_include(arch, include_filename, magic=[]):
+    if not magic:
+        magic.extend([fmt.format(arch) for fmt in INCLUDE_LOOKUP_PATHS_FORMAT])
+
+    for possible_prefix in magic:
+        possible_full_filename = possible_prefix + include_filename
+        if os.path.exists(possible_full_filename):
+            return possible_full_filename
+    return None
+
+
+def resolve_local_path(filename, relative_path):
+    filename = os.path.dirname(filename)
+
+    # Collapse '..' into the path of `filename`.
+    # Note that this is currently not used, and doesn't work in some edgecases.
+    while relative_path.startswith('../') and filename:
+        relative_path = relative_path[len('../'):]
+        filename = os.path.dirname(filename)
```

Comment:
Use https://docs.python.org/3/library/pathlib.html

Reply:
Why? os.path is also a standard library, and does exactly what's needed. In fact, as the comment explains: That case currently doesn't happen, so I'll replace the code by an `assert '../' not in filename`